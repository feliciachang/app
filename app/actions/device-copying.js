import _ from 'lodash';

import * as Types from './types';

export function copyFromDevices(devices) {
    return {
        type: Types.COPY_DEVICE_FILES,
        devices: devices
    };
}

