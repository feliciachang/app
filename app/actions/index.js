'use strict';

import * as ActionTypes from './types';

export * from './navigation';
export * from './emails';
export * from './device-status';
export * from './device-data';
export * from './timers';
export * from './modules';
export * from './configuration';
export * from './live-data';

export const Types = ActionTypes;
