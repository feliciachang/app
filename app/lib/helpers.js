'use strict';

export function unixNow() {
    return Math.round((new Date()).getTime() / 1000);
}
