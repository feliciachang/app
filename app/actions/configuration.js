import { CALL_DEVICE_API } from '../middleware/device-api';
import { QueryType } from '../lib/protocol';

import * as Types from './types';

export function queryConfiguration() {
    return {
        [CALL_DEVICE_API]: {
            types: [Types.DEVICE_NETWORK_CONFIGURATION_START, Types.DEVICE_NETWORK_CONFIGURATION_SUCCESS, Types.DEVICE_NETWORK_CONFIGURATION_FAIL],
            blocking: true,
            message: {
                type: QueryType.values.QUERY_NETWORK_SETTINGS,
            }
        }
    };
}

export function saveNetworkConfiguration(newConfiguration) {
    return {
        [CALL_DEVICE_API]: {
            types: [Types.DEVICE_SAVE_NETWORK_CONFIGURATION_START, Types.DEVICE_SAVE_NETWORK_CONFIGURATION_SUCCESS, Types.DEVICE_SAVE_NETWORK_CONFIGURATION_FAIL],
            blocking: true,
            message: {
                type: QueryType.values.QUERY_CONFIGURE_NETWORK_SETTINGS,
                networkSettings: newConfiguration,
            }
        }
    };
}

