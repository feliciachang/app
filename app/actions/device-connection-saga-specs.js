import 'react-native';

import SagaTester from 'redux-saga-tester';
import Promise from "bluebird";

import * as Types from './types';
import { findDeviceInfo } from './discovery';
import { discoverDevices, pingConnectedDevice } from './sagas';
import { useFakeDeviceConnection } from '../middleware/device-api';
import { QueryType, ReplyType } from '../lib/protocol';

describe('device connection saga', () => {
    let tester;
    let fakeDevice;

    beforeEach(() => {
        fakeDevice = useFakeDeviceConnection();
        tester = new SagaTester({
            initialState: {
                devices: { },
                deviceStatus: {
                    addresses: { },
                    connected: null,
                    api: {
                        pending: false
                    }
                }
            }
        });
    });

    describe('discovery', () => {
        let task;

        beforeEach(() => {
            task = tester.start(discoverDevices);
        });

        it.skip('should timeout and fail after timeout interval', async () => {
            await tester.waitFor(Types.FIND_DEVICE_FAIL);
        });

        it.skip('should timeout and fail when discovery returns invalid addresses', async () => {
            tester.dispatch(findDeviceInfo('127.0.0.1', 0));
            await tester.waitFor(Types.FIND_DEVICE_FAIL);
        });

        it('should query the device for capabilities and return success', async () => {
            fakeDevice.push({}, {
                type: ReplyType.values.REPLY_CAPABILITIES,
                response: { capabilities: { } }
            })
            tester.dispatch(findDeviceInfo('127.0.0.1', 12345));
            await tester.waitFor(Types.FIND_DEVICE_SUCCESS);
            expect(fakeDevice.queue).toHaveLength(0);
        });

        afterEach(() => {
            task.cancel();
        });
    });

    describe('pinging', () => {
        let task;

        beforeEach(() => {
            task = tester.start(pingConnectedDevice);
        });

        afterEach(() => {
            // Cancel to keep from interfering with the following test.
            task.cancel();
        });

        it('should ping after selecting device', () => {
            tester.getState().deviceStatus.connected = {};

            fakeDevice.push({}, {
                type: ReplyType.values.REPLY_CAPABILITIES,
                response: { capabilities: { } }
            });

            tester.dispatch({
                type: Types.FIND_DEVICE_SELECT,
                address: {}
            });

            return Promise.delay(Config.pingDeviceInterval + 100).then(() => {
                expect(fakeDevice.queue).toHaveLength(0);
            });
        });

        it('should ping after a previous ping', () => {
            tester.getState().deviceStatus.connected = {};

            fakeDevice.push({}, {
                type: ReplyType.values.REPLY_CAPABILITIES,
                response: { capabilities: { } }
            });

            tester.dispatch({
                type: Types.DEVICE_PING_SUCCESS,
                address: {}
            });

            return Promise.delay(Config.pingDeviceInterval + 100).then(() => {
                expect(fakeDevice.queue).toHaveLength(0);
            });
        });

        it('should dispatch FIND_DEVICE_LOST when ping fails', async () => {
            tester.getState().deviceStatus.connected = {};

            fakeDevice.pushError({}, "No reply");

            tester.dispatch({
                type: Types.FIND_DEVICE_SELECT,
                address: {}
            });

            await tester.waitFor(Types.FIND_DEVICE_LOST);
        });
    });
});
