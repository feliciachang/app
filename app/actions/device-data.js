'use strict';

import * as Types from './types';
import { CALL_DEVICE_API } from '../middleware/device-api';

import { QueryType } from '../lib/protocol';

export function queryCapabilities() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_CAPABILITIES_START, Types.DEVICE_CAPABILITIES_SUCCESS, Types.DEVICE_CAPABILITIES_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_CAPABILITIES,
                }
            },
        });
    };
}

export function queryInfo() {
    return (dispatch, getState) => {
        dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_CAPABILITIES_START, Types.DEVICE_CAPABILITIES_SUCCESS, Types.DEVICE_CAPABILITIES_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_CAPABILITIES,
                }
            },
        });

        dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_STATUS_START, Types.DEVICE_STATUS_SUCCESS, Types.DEVICE_STATUS_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_STATUS,
                }
            },
        });
    };
}

export function queryConfiguration() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_NETWORK_CONFIGURATION_START, Types.DEVICE_NETWORK_CONFIGURATION_SUCCESS, Types.DEVICE_NETWORK_CONFIGURATION_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_NETWORK_SETTINGS,
                }
            },
        });
    };
}

export function saveNetworkConfiguration(newConfiguration) {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_SAVE_NETWORK_CONFIGURATION_START, Types.DEVICE_SAVE_NETWORK_CONFIGURATION_SUCCESS, Types.DEVICE_SAVE_NETWORK_CONFIGURATION_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_CONFIGURE_NETWORK_SETTINGS,
                    networkSettings: newConfiguration,
                }
            },
        });
    };
}

export function queryFiles() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_FILES_START, Types.DEVICE_FILES_SUCCESS, Types.DEVICE_FILES_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_FILES
                }
            },
        });
    };
}

export function queryDataSets() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_DATA_SETS_START, Types.DEVICE_DATA_SETS_SUCCESS, Types.DEVICE_DATA_SETS_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_DATA_SETS,
                    queryDataSets: {}
                }
            },
        });
    };
}

export function queryDataSet(id) {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_DATA_SET_START, Types.DEVICE_DATA_SET_SUCCESS, Types.DEVICE_DATA_SET_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_DATA_SET,
                    queryDataSet: {
                        id: id
                    }
                }
            },
        });
    };
}

export function eraseDataSet(id) {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_ERASE_DATA_SET_START, Types.DEVICE_ERASE_DATA_SET_SUCCESS, Types.DEVICE_ERASE_DATA_SET_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_ERASE_DATA_SET,
                    eraseDataSet: {
                        id: id
                    }
                }
            },
        });
    };
}

export function startDownloadDataSet(id) {
    return (dispatch, getState) => {
        return dispatch({
            type: Types.DOWNLOAD_DATA_SET_START,
            id: id,
        });
    };
}

export function startLiveDataPoll() {
    return (dispatch, getState) => {
        return dispatch({
            type: Types.LIVE_DATA_POLL_START
        });
    };
}

export function stopLiveDataPoll() {
    return (dispatch, getState) => {
        return dispatch({
            type: Types.LIVE_DATA_POLL_STOP
        });
    };
}

export function deviceModuleQuery(id, address, message) {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_MODULE_QUERY_START, Types.DEVICE_MODULE_QUERY_SUCCESS, Types.DEVICE_MODULE_QUERY_FAIL],
                address: getState().deviceStatus.connected,
                blocking: true,
                message: {
                    type: QueryType.values.QUERY_MODULE,
                    module: {
                        id: id,
                        address: address,
                        message: message
                    }
                }
            },
        });
    };
}
