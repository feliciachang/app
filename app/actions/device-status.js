'use strict';

import _ from 'lodash';
import ServiceDiscovery from "react-native-service-discovery";
import { Alert, ToastAndroid } from 'react-native';

import * as Types from './types';
import { CALL_DEVICE_API } from '../middleware/device-api';
import { unixNow } from '../lib/helpers';

import {
    navigateWelcome,
    navigateDeviceMenu
} from './navigation';

import {
    QueryType
} from '../lib/protocol';

import Mailer from 'react-native-mail';

export function devicePing() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_PING_START, Types.DEVICE_PING_SUCCESS, Types.DEVICE_PING_FAIL],
                address: getState().deviceStatus.address,
                message: {
                    type: QueryType.values.QUERY_CAPABILITIES,
                    queryCapabilities: {
                        version: 1
                    }
                }
            },
        });
    };
}

export function queryDeviceCapabilities() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_CAPABILITIES_START, Types.DEVICE_CAPABILITIES_SUCCESS, Types.DEVICE_CAPABILITIES_FAIL],
                address: getState().deviceStatus.address,
                message: {
                    type: QueryType.values.QUERY_CAPABILITIES,
                    queryCapabilities: {
                        version: 1
                    }
                }
            },
        });
    };
}

export function queryDataSet(id) {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_DATA_SET_START, Types.DEVICE_DATA_SET_SUCCESS, Types.DEVICE_DATA_SET_FAIL],
                address: getState().deviceStatus.address,
                message: {
                    type: QueryType.values.QUERY_DATA_SET,
                    queryDataSet: {
                        id: id
                    }
                }
            },
        });
    };
}

export function eraseDataSet(id) {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_ERASE_DATA_SET_START, Types.DEVICE_ERASE_DATA_SET_SUCCESS, Types.DEVICE_ERASE_DATA_SET_FAIL],
                address: getState().deviceStatus.address,
                message: {
                    type: QueryType.values.QUERY_ERASE_DATA_SET,
                    eraseDataSet: {
                        id: id
                    }
                }
            },
        });
    };
}

export function emailDataSet(id) {
    return (dispatch, getState) => {
        Mailer.mail({
            subject: 'FieldKit NOAA-CTD Data',
            recipients: ['jlewalle@gmail.com'],
            body: '<p>Please see the attached file, data.csv.</p><br/><p>Thanks!</p>',
            isHTML: true,
            /*
            attachment: {
                path: '',
                type: 'text/csv',
                name: 'data.csv',
            }
            */
        }, (error, event) => {
            Alert.alert(
                error,
                event,
                [
                    {text: 'Ok', onPress: () => console.log('OK: Email Error Response')},
                    {text: 'Cancel', onPress: () => console.log('CANCEL: Email Error Response')}
                ],
                { cancelable: true }
            )

            dispatch({
                type: Types.EMAIL_DATA_SET_SUCCESS,
                id: id
            });

            ToastAndroid.show('E-mail sent successfully.', ToastAndroid.SHORT);
        });
    };
}

export function queryDataSets() {
    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: [Types.DEVICE_DATA_SETS_START, Types.DEVICE_DATA_SETS_SUCCESS, Types.DEVICE_DATA_SETS_FAIL],
                address: getState().deviceStatus.address,
                message: {
                    type: QueryType.values.QUERY_DATA_SETS,
                    queryDataSets: {}
                }
            },
        });
    };
}

let serviceDiscovery = null;
let pingTimer = null;

function tick() {
    return (dispatch, getState) => {
        const { nav, deviceStatus } = getState();
        const route = nav.routes[nav.index];

        if (deviceStatus.address.valid) {
            if (!deviceStatus.api.pending) {
                if (unixNow() - deviceStatus.ping.time > 30) {
                    devicePing()(dispatch, getState);
                }
                else {
                    if (deviceStatus.ping.success) {
                        if (_.isObject(route.params) && route.params.connecting === true) {
                            dispatch(queryDeviceCapabilities());
                            dispatch(navigateDeviceMenu());
                        }
                    }
                    else {
                        if (_.isObject(route.params) && route.params.connectionRequired === true) {
                            dispatch(navigateWelcome());
                        }
                    }
                }
            }
            else {

            }
        }
        else {
            if (unixNow() - deviceStatus.started > 60) {
                if (_.isObject(route.params) && route.params.connecting === true) {
                    dispatch(navigateWelcome());
                }
            }
        }
    };
}

export function deviceStartConnect() {
    return (dispatch, getState) => {
        dispatch({
            type: Types.DEVICE_CONNECT_START,
        });

        if (serviceDiscovery === null) {
            serviceDiscovery = new ServiceDiscovery();

            serviceDiscovery.on('service-resolved', (ev) => {
            });

            serviceDiscovery.on('udp-discovery', (ev) => {
                const oldAddress = getState().deviceStatus.address;
                const address = {
                    host: ev.address,
                    port: ev.port,
                    valid: true
                };
                if (!_.isEqual(oldAddress, address)) {
                    dispatch({
                        type: Types.DEVICE_CONNECT_INFO,
                        address: address
                    });
                }
            });
        }

        serviceDiscovery.start();

        clearInterval(pingTimer);
        pingTimer = setInterval(() => {
            tick()(dispatch, getState);
        }, 1000);
    };
}

export function deviceStopConnect() {
    return (dispatch, getState) => {
        if (pingTimer != null) {
            clearInterval(pingTimer);
            pingTimer = null;
        }

        dispatch({
            type: Types.DEVICE_CONNECT_STOP,
        });
    };
}
