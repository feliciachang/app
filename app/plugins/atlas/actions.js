import { delay } from 'redux-saga';
import { put, take, takeLatest, takeEvery, select, all, race, call } from 'redux-saga/effects';

import * as AppActionTypes from '../../actions/types';
import { QueryType as AppQueryType } from '../../lib/protocol';

import * as AtlasActionTypes from './types';
import { deviceCall } from '../../actions/saga-utils';
import { CALL_DEVICE_API } from '../../middleware/device-api';

import { QueryType, SensorType, atlasSensorQuery, encodeWireAtlasQuery, decodeWireAtlasReply } from './protocol';

export function atlasCalibrationBegin(sensor) {
    return {
        type: AtlasActionTypes.ATLAS_CALIBRATION_BEGIN,
        sensor: sensor,
    };
}

export function atlasCalibrationEnd() {
    return {
        type: AtlasActionTypes.ATLAS_CALIBRATION_END,
    };
}

export function atlasReadSensor(sensor) {
    return {
        type: AtlasActionTypes.ATLAS_READ_SENSOR_START,
        sensor: sensor,
    };
}

export function atlasSensorCommand(types, blocking, sensor, command) {
    const query = atlasSensorQuery(sensor, command);
    const encoded = encodeWireAtlasQuery(query);

    return (dispatch, getState) => {
        return dispatch({
            [CALL_DEVICE_API]: {
                types: types,
                address: getState().deviceStatus.connected,
                blocking: blocking,
                message: {
                    type: AppQueryType.values.QUERY_MODULE,
                    module: {
                        id: 8,
                        address: 8,
                        message: encoded
                    }
                }
            },
        });
    };
}

export function atlasSetProbeType(sensor, command) {
    const types = [AtlasActionTypes.DEVICE_ATLAS_SENSOR_SET_PROBE_TYPE_START, AtlasActionTypes.DEVICE_ATLAS_SENSOR_SET_PROBE_TYPE_SUCCESS, AtlasActionTypes.DEVICE_ATLAS_SENSOR_SET_PROBE_TYPE_FAIL];
    return atlasSensorCommand(types, true, sensor, command);
}

export function atlasCalibrate(sensor, command) {
    const types = [AtlasActionTypes.DEVICE_ATLAS_SENSOR_CALIBRATE_START, AtlasActionTypes.DEVICE_ATLAS_SENSOR_CALIBRATE_SUCCESS, AtlasActionTypes.DEVICE_ATLAS_SENSOR_CALIBRATE_FAIL];
    return atlasSensorCommand(types, true, sensor, command);
}

export function atlasReading(sensor) {
    const types = [AtlasActionTypes.DEVICE_ATLAS_SENSOR_READING_START, AtlasActionTypes.DEVICE_ATLAS_SENSOR_READING_SUCCESS, AtlasActionTypes.DEVICE_ATLAS_SENSOR_READING_FAIL];
    return atlasSensorCommand(types, false, sensor, "R");
}

function* takeAtlasReadings() {
    yield takeLatest([AtlasActionTypes.ATLAS_READ_SENSOR_START], function* (start) {
        while (true) {
            yield call(deviceCall, atlasReading(start.sensor));

            const { stopped, to } = yield race({
                stopped: take([AtlasActionTypes.ATLAS_READ_SENSOR_STOP, AtlasActionTypes.ATLAS_CALIBRATION_END]),
                to: delay(2000)
            });

            if (stopped) {
                break;
            }
        }
    });
}

export function* atlasSagas() {
    yield all([
        takeAtlasReadings()
    ]);
}
