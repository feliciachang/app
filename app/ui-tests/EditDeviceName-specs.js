import React from "react";
import EditDeviceName from "../containers/EditDeviceNameScreen";
import { mockFunctionHelper } from "../../specs/helpers";

import { View, Button } from "react-native";
import { shallow, mount, render } from "enzyme";
import toJson from "enzyme-to-json";
import * as Types from "../actions/index";

let store = mockFunctionHelper({
    nav: {
        index: 1,
        routes: [
            {
                key: "Init-id-1560200089303-0",
                routeName: "EasyModeWelcome",
                type: undefined
            },
            {
                key: "id-1560200089303-1",
                params: {
                    address: {
                        host: "192.168.2.1",
                        key: "192.168.2.1",
                        port: 54321,
                        valid: true
                    },
                    deviceId: "0004a30b0023308d"
                },
                routeName: "EditDeviceName"
            }
        ]
    }
});

describe.only("Set Device Name Button", () => {
    it("simulates click event", () => {
        const wrapper = shallow(<EditDeviceName store={store} />);
        // expect(wrapper.find(View).length).toEqual(20);
        expect(wrapper.find(Button).length).toEqual(2);
    });
});

describe("<EditDeviceName/>", () => {
    describe("render()", () => {
        test("renders the component", () => {
            const wrapper = shallow(<EditDeviceName store={store} />);
            const component = wrapper.dive();
            expect(toJson(component)).toMatchSnapshot();
        });
    });
});
