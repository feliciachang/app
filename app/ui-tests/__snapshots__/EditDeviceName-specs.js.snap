// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditDeviceName/> render() renders the component 1`] = `
<Component>
  <Component
    style={
      Object {
        "paddingBottom": 10,
        "paddingTop": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Device Name"
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "height": 40,
        }
      }
      underlineColorAndroid="transparent"
    />
  </Component>
  <Component
    style={
      Object {
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Button
      onPress={[Function]}
      title="Save"
    />
  </Component>
  <Component
    style={
      Object {
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <Button
      onPress={[Function]}
      title="Cancel"
    />
  </Component>
</Component>
`;
